File: memex/services.py
"""Manage IPFS daemon lifecycle and dagit identity."""

import subprocess
import time
from pathlib import Path

import httpx


def is_ipfs_running() -> bool:
    """Check if IPFS daemon is already running."""
    try:
        resp = httpx.post("http://localhost:5001/api/v0/id", timeout=2)
        return resp.status_code == 200
    except httpx.RequestError:
        return False


def ensure_ipfs_repo(ipfs_bin: str) -> None:
    """Initialize IPFS repo if it doesn't exist."""
    if (Path.home() / ".ipfs").exists():
        return

    print("\033[0;32m[memex]\033[0m Initializing IPFS repository...")
    result = subprocess.run(
        [ipfs_bin, "init"],
        capture_output=True,
        text=True,
    )
    if result.returncode != 0:
        print(f"\033[0;31m[memex]\033[0m IPFS init failed: {result.stderr}")
        raise SystemExit(1)
    print("\033[0;32m[memex]\033[0m IPFS repository initialized")


def start_ipfs_daemon(ipfs_bin: str) -> subprocess.Popen:
    """Start IPFS daemon as a background process."""
    print("\033[0;32m[memex]\033[0m Starting IPFS daemon...")
    proc = subprocess.Popen(
        [ipfs_bin, "daemon"],
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
    )
    return proc


def wait_for_ipfs(timeout: float = 30) -> bool:
    """Wait for IPFS daemon API to become available."""
    deadline = time.time() + timeout
    while time.time() < deadline:
        if is_ipfs_running():
            return True
        time.sleep(0.5)
    return False


def ensure_dagit_identity() -> str:
    """Create dagit identity if missing, return DID.

    Returns:
        The user's DID string.
    """
    identity_path = Path.home() / ".dagit" / "identity.json"

    if identity_path.exists():
        import json
        data = json.loads(identity_path.read_text())
        return data.get("did", "unknown")

    print("\033[0;32m[memex]\033[0m Creating dagit identity...")
    try:
        from dagit.identity import create
        identity = create()
        did = identity.get("did", "unknown") if isinstance(identity, dict) else str(identity)
        print(f"\033[0;32m[memex]\033[0m Identity created: {did}")
        return did
    except Exception as e:
        print(f"\033[1;33m[memex]\033[0m Could not create dagit identity: {e}")
        return "unknown"

File: src/services.ts
/**
 * Subprocess management, health checks, and cleanup.
 */

import { existsSync, mkdirSync, readdirSync } from "fs";
import { join } from "path";
import { homedir } from "os";
import type { Subprocess } from "bun";

const procs: Subprocess[] = [];
let cleaningUp = false;

function log(msg: string) {
  console.log(`\x1b[0;32m[memex]\x1b[0m ${msg}`);
}

function warn(msg: string) {
  console.log(`\x1b[1;33m[memex]\x1b[0m ${msg}`);
}

function error(msg: string) {
  console.error(`\x1b[0;31m[memex]\x1b[0m ${msg}`);
}

export async function isIpfsRunning(): Promise<boolean> {
  try {
    const resp = await fetch("http://localhost:5001/api/v0/id", {
      method: "POST",
      signal: AbortSignal.timeout(2000),
    });
    return resp.status === 200;
  } catch {
    return false;
  }
}

export async function ensureIpfsRepo(ipfsBin: string): Promise<void> {
  if (existsSync(join(homedir(), ".ipfs"))) return;

  log("Initializing IPFS repository...");
  const proc = Bun.spawn([ipfsBin, "init"], {
    stdout: "pipe",
    stderr: "pipe",
  });
  await proc.exited;

  if (proc.exitCode !== 0) {
    const stderr = await new Response(proc.stderr).text();
    error(`IPFS init failed: ${stderr}`);
    process.exit(1);
  }
  log("IPFS repository initialized");
}

export async function startIpfsDaemon(ipfsBin: string): Promise<Subprocess> {
  log("Starting IPFS daemon...");
  const proc = Bun.spawn([ipfsBin, "daemon"], {
    stdout: "ignore",
    stderr: "ignore",
  });
  procs.push(proc);
  return proc;
}

export async function waitForIpfs(timeout = 30000): Promise<boolean> {
  const deadline = Date.now() + timeout;
  while (Date.now() < deadline) {
    if (await isIpfsRunning()) return true;
    await Bun.sleep(500);
  }
  return false;
}

export async function ensureIdentity(): Promise<string> {
  const { loadIdentity, createIdentity } = await import("./identity");

  const existing = await loadIdentity();
  if (existing) return existing.did ?? "unknown";

  log("Creating identity...");
  try {
    const identity = await createIdentity();
    log(`Identity created: ${identity.did}`);
    return identity.did;
  } catch (e: any) {
    warn(`Could not create identity: ${e.message}`);
    return "unknown";
  }
}

export function isMounted(mountPath: string): boolean {
  try {
    const entries = readdirSync(mountPath);
    // A live memex-fs mount has nodes/, types/, search/ at minimum
    return entries.includes("nodes") && entries.includes("types") && entries.includes("search");
  } catch {
    return false;
  }
}

export function isGraphEmpty(mountPath: string): boolean {
  try {
    const entries = readdirSync(join(mountPath, "nodes"));
    return entries.length === 0;
  } catch {
    return true;
  }
}

export function cleanupAll(): void {
  if (cleaningUp) return;
  cleaningUp = true;

  for (const proc of procs.reverse()) {
    try {
      proc.kill("SIGTERM");
      // Give 5s then force kill
      setTimeout(() => {
        try { proc.kill("SIGKILL"); } catch {}
      }, 5000);
    } catch {}
  }
}

export function registerCleanup(): void {
  process.on("SIGINT", () => {
    process.stdout.write("\x1b[?1049l"); // exit alt screen
    cleanupAll();
    process.exit(0);
  });
  process.on("SIGTERM", () => {
    process.stdout.write("\x1b[?1049l"); // exit alt screen
    cleanupAll();
    process.exit(0);
  });
}

export { log, warn, error };
